### PR TITLE
fix: update to newer ec-gpu version

### PR DESCRIPTION
**Pull request by vmx**
The API of ec-gpu changed. The new version makes naming things easier,
as it can directly be derived from the fields. Also helper functions
where introduced so that less boilerplate is needed for generating
a GPU kernel.

A new trait alias called `NeptuneField` is introduced. It changes the
traits, depending on whether GPU support is enabled or not (via the
`cuda`/`opencl` features). This makes the code easier to read and
maintain. Without that trait alias at places like this:

    impl<F, A> Batcher<F, A>
    where
        F: NeptuneField,
        A: Arity<F>,

You would instead need to write:

    impl<
            #[cfg(not(any(feature = "cuda", feature = "opencl")))] F: PrimeField,
            #[cfg(any(feature = "cuda", feature = "opencl"))] F: PrimeField + GpuField,
            A: Arity<F>,
        > Batcher<F, A>

Setting trait bounds with conditional compilation is not supported
in `where` clauses.

This version uses a fork of `pasta_curves`. Once upstream `pasta_curves`
v0.5.0 is released, it will be switched again. This temporary for is
only needed in order to make things work with the `ec-gpu` v0.2.

BREAKING CHANGE: deeper down dependencies changed

While this change itself is not really a breaking change, updating
to the newer `blstrs` version is. The reason is that it depends on
a pinned version of `blst`, which was updated. If anything within
your dependency tree is using a different version of `blstrs` you
will end up hard to resolve dependency issues.

BREAKING CHANGE: environment variables are renamed

`NEPTUNE_GPU_FRAMEWORK` is now called `EC_GPU_FRAMEWORK`,
`NEPTUNE_CUDA_NVCC_ARGS` is now called `EC_GPU_CUDA_NVCC_ARGS`.

BREAKING CHANGE: the `wasm` feature was removed

Due to the upgrade to a newer `bellperson` release, the `wasm` feature
is not longer needed. Compiling for WebAssembly should work out of the
box.

---

This PR is not ready to be merged as it relies on https://github.com/zcash/pasta_curves/pull/45, but it can already be reviewed in case any changes are needed.

### CODE REVIEW

**Comment by vmx** [comment]
Does anyone have the time to review those changes?

**Comment by porcuquine** [comment]
I don't right now, but maybe next week — especially if you remind me.

**Comment by vmx** [comment]
> especially if you remind me.

I hereby remind you :D

File: src/lib.rs [hunk 1]
```diff
@@ -74,6 +75,23 @@ pub mod batch_hasher;
 #[cfg(any(feature = "cuda", feature = "opencl"))]
 pub mod proteus;
 
+#[cfg(not(any(feature = "cuda", feature = "opencl")))]
+trait_set! {
```
Comment (porcuquine) [inline src/lib.rs:79]: Does this create a new dependency on unstable Rust?

**Review by porcuquine — APPROVED** [review]
Noting that @vmx says this isn't actually ready to merge yet, and apart from one comment questioning whether use of the `trait_set!` macro may cause problems, this looks fine to me.

File: src/lib.rs (hunk 1 above)
Comment (vmx) [reply src/lib.rs:79]: The reason to use the `trait_set` macro is that this way we do *not* need unstable Rust.

**Comment by vmx** [comment]
I did a force-push as the difference is mostly getting rid of the `patch` part: https://github.com/filecoin-project/neptune/compare/4f385b4f678f3b2366e2675a9eb76cb533cfe2c2..650e43231d159ab36ce710475be0d5ab9edc7f62. I've also updated the commit message.

Using a fork for `pasta_curves` is not ideal and I had hoped [the change I needed](https://github.com/zcash/pasta_curves/pull/45) would get merged and released earlier, but it didn't so far. Hence I propose using a (short-lived) fork for now, which is the upstream 0.4.1 release plus that single change. Ideally upgrading to the once released upstream version should be smooth.

**Pull request merged**